tools/lib/report/report.go

```diff
@@ -121,17 +121,31 @@ func asBin(report *spb.Attestation) ([]byte, error) {
 	return append(r, certs...), nil
 }
 
-func tcbBreakdown(tcb uint64) string {
-	parts := kds.DecomposeTCBVersion(kds.TCBVersion(tcb))
-	return fmt.Sprintf("0x%x:{ucode: %d, snp: %d, tee: %d, bl: %d}", tcb, parts.UcodeSpl, parts.SnpSpl,
-		parts.TeeSpl, parts.BlSpl)
+func tcbBreakdown(tcbVersion uint8, tcb uint64) string {
+	parts := kds.DecomposeTCBVersion(kds.TCBVersion{Version: tcbVersion, TCB: tcb})
+	switch tcbVersion {
+	case kds.TCBStructVersion1:
+		return fmt.Sprintf("0x%x:{ucode: %d, snp: %d, tee: %d, bl: %d, fmc: %d}",
+			tcb, parts.UcodeSpl, parts.SnpSpl, parts.TeeSpl, parts.BlSpl, parts.FmcSpl)
+	case kds.TCBStructVersion0:
+		fallthrough
+	default:
+		return fmt.Sprintf("0x%x:{ucode: %d, snp: %d, tee: %d, bl: %d}",
+			tcb, parts.UcodeSpl, parts.SnpSpl, parts.TeeSpl, parts.BlSpl)
+	}
 }
 
 func tcbText(report *spb.Attestation) ([]byte, error) {
+	fms := report.GetReport().GetCpuid1EaxFms()
+	tcbVersion, err := kds.ProductLineToTCBVersion(kds.ProductLineFromFms(fms))
+	if err != nil {
+		return nil, err
+	}
+
 	return []byte(fmt.Sprintf("current_tcb=%s\ncommitted_tcb=%s\nlaunch_tcb=%s\n",
-		tcbBreakdown(report.Report.GetCurrentTcb()),
-		tcbBreakdown(report.Report.GetCommittedTcb()),
-		tcbBreakdown(report.Report.GetLaunchTcb()))), nil
+		tcbBreakdown(tcbVersion, report.Report.GetCurrentTcb()),
+		tcbBreakdown(tcbVersion, report.Report.GetCommittedTcb()),
+		tcbBreakdown(tcbVersion, report.Report.GetLaunchTcb()))), nil
```

Comment:
Instead of passing this multiple times, would it make sense to call `ProductLineToTCBVersion` inside of tcbBreakdown?

Reply:
We anyway have to call tcbBreakdown for Current, Committed and Launch TCBs.

I don't have a preference, we can do it either way. If we do  `ProductLineToTCBVersion` inside `tcbBreakdown`, then we have to pass the product name to it instead of tcbVersion.

Reply:
@larrydewey I'll mark this conversation as resolved if you don't have objection to the existing approach. Please re-open it if you want to discuss this further. Thank you!